### PR TITLE
[JENKINS-69036] Do not always log errors if input is empty 

### PR DESCRIPTION
**Pull request by shamlymhd**
Do not log an error if input is empty for those parsers that can read from the console log: in those cases an empty input just means that there are no warnings, see [JENKINS-69036](https://issues.jenkins.io/browse/JENKINS-69036) for details.

### CODE REVIEW

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java [hunk 1]
```diff
@@ -110,6 +110,9 @@ public boolean getSkipSymbolicLinks() {
     private boolean followSymlinks() {
         return !getSkipSymbolicLinks();
     }
+    public boolean errorOnEmptyFiles() {
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java:113]:
```suggestion
    private boolean isEmptyFileValid() {
```

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java [hunk 2]
```diff
@@ -166,7 +169,7 @@ private String expandPattern(final Run<?, ?> run, final String actualPattern) {
     private Report scanInWorkspace(final FilePath workspace, final String expandedPattern, final LogHandler logger) {
         try {
             FileVisitorResult<Report> report = workspace.act(
-                    new IssueReportScanner(expandedPattern, reportEncoding, followSymlinks(), createParser()));
+                    new IssueReportScanner(expandedPattern, reportEncoding, followSymlinks(), createParser(), errorOnEmptyFiles()));
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java:172]:
```suggestion
                    new IssueReportScanner(expandedPattern, reportEncoding, followSymlinks(), createParser(), !isEmptyFileValid()));
```

**Review by uhafner — COMMENTED** [review]
It would be also have two tests, one for a parser that scans the console log and one for an XML parser. See bug report for the former one.

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java (hunk 1 above)
Comment (shamlymhd) [reply plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java:113]: Resolved

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java (hunk 2 above)
Comment (shamlymhd) [reply plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java:172]: Resolved

**Comment by shamlymhd** [comment]
See bug report for the former one. -> All SpotBugs are empty. so where should I look for?

**Comment by uhafner** [comment]
> This issue is closed. right?

Not yet. Each PR requires a test case, see my comments above.

**Comment by uhafner** [comment]
Here is an example test case:

https://github.com/jenkinsci/warnings-ng-plugin/blob/master/plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java#L489

This one fails, when there are no files found. You need to create a similar one that reads an empty file. 

When you enable your new option, the new test should not fail for Oxygen or JavaDoc. For a tool like PMD, CheckStyle, that read XML, there still should be an error when the input is empty.

**Comment by uhafner** [comment]
@shamlymhd ping! Are you planning to finish this PR?

**Comment by shamlymhd** [comment]
@uhafner yeah I had exams in the last two weeks and am back to this today.

**Comment by uhafner** [comment]
Ok, no worries! I just wasn't sure if you are waiting for some input. See take all the time you need!

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java [hunk 3]
```diff
@@ -498,6 +499,33 @@ void shouldFilterByMessage() {
         assertThat(result.getIssues()).hasSize(3);
     }
 
+    /** Runs the JavaDoc parser on empty file , shoud not report an error */
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java:502]:
```suggestion
    /** Verifies that the JavaDoc parser reports just an information message when reading an empty file. */
```

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java [hunk 4]
```diff
@@ -498,6 +499,33 @@ void shouldFilterByMessage() {
...
+        job.setDefinition(asStage(
+                "recordIssues tool: javaDoc(pattern:'**/*issues.txt', reportEncoding:'UTF-8')"));
+
+        AnalysisResult result = scheduleSuccessfulBuild(job);
+        assertThat(result.getIssues()).hasSize(0);
+        assertThat(result).hasInfoMessages(
+                "Skipping file 'emptyFile-issues.txt' because it's empty");
+    }
+
+    /** Runs the checkStyle parser that read XML on empty file , shoud report an error */
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java:516]:
```suggestion
    /** Verifies that the CheckStyle parser reports an error when reading an empty XML file. */
```

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java [hunk 5]
```diff
@@ -498,6 +499,33 @@ void shouldFilterByMessage() {
         assertThat(result.getIssues()).hasSize(3);
     }
 
+    /** Runs the JavaDoc parser on empty file , shoud not report an error */
+    @Test
+    void shouldNotReportError() {
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java:504]:
```suggestion
    void javaDocShouldNotReportErrorOnEmptyFiles() {
```

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java [hunk 6]
```diff
@@ -498,6 +499,33 @@ void shouldFilterByMessage() {
...
+
+        AnalysisResult result = scheduleSuccessfulBuild(job);
+        assertThat(result.getIssues()).hasSize(0);
+        assertThat(result).hasInfoMessages(
+                "Skipping file 'emptyFile-issues.txt' because it's empty");
+    }
+
+    /** Runs the checkStyle parser that read XML on empty file , shoud report an error */
+    @Test
+    void shouldReportError() {
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java:518]:
```suggestion
    void checkStyleShouldReportErrorOnEmptyFiles() {
```

**Review by uhafner — APPROVED** [review]
Just some small documentation improvements.

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java [hunk 7]
```diff
@@ -30,6 +30,7 @@
 import org.jenkinsci.plugins.workflow.graphanalysis.DepthFirstScanner;
 import org.jenkinsci.plugins.workflow.job.WorkflowJob;
 import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import hudson.model.FreeStyleProject;
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java:33]:
```suggestion
```

**Pull request merged**